openupgradelib/openupgrade.py

```diff
@@ -979,17 +1091,15 @@ def logged_query(cr, query, args=None, skip_no_result=False):
     :param skip_no_result: If True, then logging details are only shown
       if there are affected records.
     """
-    if args is None:
-        args = ()
     args = tuple(args) if type(args) == list else args
     try:
         cr.execute(query, args)
     except (ProgrammingError, IntegrityError):
-        logger.error('Error running %s' % cr.mogrify(query, args))
+        logger.error('Error running %s' % cr._obj.query.decode())
         raise
     if not skip_no_result or cr.rowcount:
-        logger.debug('Running %s', query % args)
-        logger.debug('%s rows affected', cr.rowcount)
+        logger.debug('%s rows affected after running %s',
```

Comment:
Why are you changing this base method that is working since a lot and with your changes can have side effects across all the versions that exists?

Reply:
Yes, let me explain.

I found the [`psycopg2.sql`](http://initd.org/psycopg/docs/sql.html) module with lots of great features for a SQL-composing library such as openupgradelib, and I'm using it in the new method being added.

However, you can't just do `query % args` as it was being done before, because these are sql-specific objects, not raw strings, and don't support that. BTW, that was misleading because the logged query was not exactly the query that was being sent to the database. [As explained in psycopg2 docs](http://initd.org/psycopg/docs/usage.html?highlight=never#sql-injection):

> Warning
> Never, **never, NEVER** use Python string concatenation (+) or string parameters interpolation (%) to pass variables to a SQL query string. Not even at gunpoint. 

A few lines above there was another logger that did it more accurately: `logger.error('Error running %s' % cr.mogrify(query, args))`. This returns the same string that is sent to Postgres server, so it's a better candidate for debugging. The problem with that one is that under Python 3 it's a `bytes()` object, which means that the log is not properly formatted (basically it prints a raw `\n` instead of a real newline character, making it very hard to read by humans), and that's why I'm using `.decode()` at the end, which works fine both in Python 2 and 3.

About why use `cr._obj.query` instead of `cr.mogrify(query, args))`, it's because it's a property already found in all cursors.

I checked it works from v8 to v12, but TBH didn't check that in v7 or lower.

Maybe I could use the mogrify + decode version, which should work fine always, regardless of wether that hidden `cr._obj.query` attribute exists or not.

Reply:
Oh and I forgot it: now that I was changing it, I didn't see a reason to use 2 log lines, so I reduced it to 1. Not a big change, though.

Reply:
You should make use to explain all of this in advance in the commit message, so it's kept for later references. OK to the change, but please include all these explanations in the commit message.

About `mogrify` you can simply check the existence of the field with `hasattr` and if not, use it as fallback.